### DISCUSSION TITLE
Hotkey Duplicate Detection

### DISCUSSION THREAD

**Discussion by dodgepong** [obsproject/obs-studio#5080] [Requests For Proposals (Bounties)]
# Duplicate Hotkey Warning in Settings

## Feature Summary

Show a warning in the hotkeys settings panel if a key or key combination is in use by another hotkey shortcut.

## Background and Motivation

Right now, OBS users are able to set multiple actions to the same shortcut. Frequently this can be intentional, as some users wish for OBS to be able to do multiple things with a single keypress. However, this is not always the case, and users may want to know when a hotkey they set is already in use so that they are fully aware of what might happen when the press that hotkey. Therefore, OBS should show a warning when a duplicate shortcut is set, but because we want to keep the ability to trigger multiple actions with the same shortcut, the warning should not be interruptive.

## Request For Proposal

Submit a proposal explaining how you will implement this feature such that it meets the following requirements:

* A visual warning is shown in the hotkeys settings panel when a hotkey is used by two or more items.
* Any text in the warning uses the Qt translated strings feature.
* The visual warning is not a pop-up/modal, nor is it interruptive. That is, it should not require user interaction to dismiss, nor should it be unnecessarily overbearing.
* The feature should work on Windows, macOS, and Linux.*

\* Multiple contributors may assist with implementation on different platforms if necessary, and may split the bounty according to the level of contribution they made to the project.

Please read [Tips for Writing a Good Proposal](https://github.com/obsproject/obs-studio/wiki/Tips-for-Writing-a-Good-Proposal) for guidance on how to author your proposal.

The OBS development team will consider all submitted proposals and select one submission to be implemented by the developer who submitted it. That developer will be eligible to collect the bounty upon completion of the project according to the criteria listed below.

## Bounty

The bounty for implementing this feature is **$1000**.

You can learn more about the OBS Project Bounty Program [here](https://github.com/obsproject/obs-studio/wiki/OBS-Project-Bounty-Program).

If you wish to contribute funds toward increasing the amount on this bounty, contribute to the [OBS Project Bounty Fund](https://opencollective.com/obs-project-bounty-program) and select the "Hotkey Duplicate Detection" option (or just [click here](https://opencollective.com/obs-project-bounty-program/contribute/bounty-hotkey-duplicate-detection-31039/checkout)).

## Completion Criteria

You may collect the bounty when the following are true:

* Your proposal addresses all items listed in the "Request For Proposal" section.
* You have submitted one more more pull requests that fully implement the feature as you proposed.
* Your pull request(s) are reviewed, and you have responded to review comments within a reasonable time frame.
* Your pull request(s) are merged into the `master` branch of the appropriate `obsproject` repositories.

## Additional Information

Initial proposal from OBS Ideas page: [https://ideas.obsproject.com/posts/1327/duplicate-hotkeys-warning](https://ideas.obsproject.com/posts/1327/duplicate-hotkeys-warning)

**Comment by furious** [comment]
I found that interesting to do, here's my proposal for this task:

- Add a new icon/button that indicates when the hotkey is duplicated (not sure which icon to use but maybe something like warning: ⚠️)
- Add a checkbox on the filter field to search by hotkey
- Clicking on the icon/button will filter using the same hotkey 

![image](https://user-images.githubusercontent.com/3678215/128403853-c001147b-d12e-4dd3-9e39-60ab2bdf8948.png)
or
![image](https://user-images.githubusercontent.com/3678215/128408320-952891c5-af6d-417d-8c25-1f173382b53d.png)

**Reply by WizardCM** [reply]
I think having a left-aligned ⚠️ warning icon that isn't mixed in with the other icons would be a more obvious UX.

**Reply by furious** [reply]
Yes, I would also add a tooltip on that field/warning icon and text on the bottom saying something like "This hotkey is being used by X actions"

**Reply by jp9000** [reply]
Sorry it's taken a while to get back to you.

I think the proposal is simple and works. What do you think @dodgepong?

**Reply by dodgepong** [reply]
Looks good to me. Like @WizardCM, I prefer the second option with the warning appearing inside the box instead of outside the box.

@furious I'm willing to accept this proposal. Are you available to work on the project for the bounty?

**Reply by furious** [reply]
@dodgepong sure! When can I start working on it? 😃

**Reply by dodgepong** [reply]
@furious Great! I will mark this bounty as In Progress, and you can start as soon as you're able. Are you on the OBS Discord? That may be an easier place to ask quick questions if you have any.

**Reply by furious** [reply]
@dodgepong okay, cool! I'm already on the OBS Discord. I'm gonna setup the dev environment to start doing something. Thanks

**Reply by furious** [reply]
Alright, I've successfully implemented the new filter by hotkey, here's some screenshots: 

![image](https://user-images.githubusercontent.com/3678215/132989011-38c7f121-a2e9-4e25-8d33-676ffc006476.png)
![image](https://user-images.githubusercontent.com/3678215/132989021-f9238131-12b4-40a7-95a8-8cbc14c46b34.png)

I'm not sure how I should scan for hotkey duplicate detection and show the warning when it finds something duplicate:

- [ ] When user change the hotkey input field (We have KeyChanged signal, easy to implement)
- [ ] When user focus/click on the hotkey input field (Not sure how to achieve it on Qt yet, any tips?)
- [ ] When user open the hotkey settings (Maybe not a good idea...)

@dodgepong @jp9000 What do you think? Also, do I need to translate the strings to all languages or just en_US is ok?

**Reply by dodgepong** [reply]
The warning indicator next to the hotkey should always be visible as long as the hotkey is shared with another. It should not depend on whether the user has selected/focused the field. Users should be able to scroll through their list of hotkeys and see any with the conflict indicator next to them. Each indicator should have a tooltip where hovering over it (or possibly clicking it) should list the hotkeys that this hotkey conflicts with.

You don't need to handle localization beyond en_US. Localization will be handled by our CrowdIn volunteers.

**Reply by furious** [reply]
Ok, now I understand better what is needed, thanks for clarifying it for me.

**Reply by furious** [reply]
Ok, here's the current state of the implementation:
![image](https://user-images.githubusercontent.com/3678215/133867348-415aaa62-d710-466f-80fc-019cfe54642c.png)

If you have any better suggestion for the tooltip text (english isn't my main language).

Is that what you guys need? Can I create the pull request so you can test?

**Reply by dodgepong** [reply]
Can you make the icon right-alighted rather than left-aligned?

Does clicking the icon just add the hotkey to the current filter?

For the tooltip, I'll defer to @Warchamp7 but my first pass would probably be something like "This hotkey is shared by one or more other actions (click to resolve):" and then a list of the names of the conflicting hotkeys.

**Reply by furious** [reply]
@dodgepong I've made it right-aligned and changed the tooltip as you suggested, what do you think?
![image](https://user-images.githubusercontent.com/3678215/133994683-ed93ec25-a1d5-4163-9bbf-a61ab45c5a42.png)

Yes, clicking in the icon will add the hotkey to the filter and only list the actions with that hotkey.
![image](https://user-images.githubusercontent.com/3678215/133995058-d7c44753-1e1e-4ef2-9cb3-4e10875360b0.png)

**Reply by dodgepong** [reply]
That looks good to me. Time to get some wider feedback on the pull request.

**Reply by Fenrirthviti** [reply]
I'm not super happy with the "Click to resolve" verbiage here, as it implies that it's doing an action to fix the problem, rather than just listing the hotkeys. Maybe something like "Click to show conflicts" instead?

Rest looks good to me.

**Reply by furious** [reply]
@Fenrirthviti I agree, translating to my native language "Click to show conflicts" seems better

**Reply by furious** [reply]
@dodgepong I've changed the text like @Fenrirthviti suggested and fixed a small issue when filtering with text and hotkey.

There's anything else I need to change?

**Reply by dodgepong** [reply]
At this point we should keep all discussion on the pull request: https://github.com/obsproject/obs-studio/pull/5297

I'd like to see a review from @Warchamp7 before moving to merge.

**Reply by dodgepong** [reply]
Congrats, your pull request has been merged! I've added you on Discord (dodgepong#8080), and will follow up with you to help you submit your invoice. Thanks so much for your contribution!

**Reply by furious** [reply]
Cool! I'm so happy to contribute to the project that I love to use it. 

I just accepted you on Discord, thank you @dodgepong!

**Comment by henke37** [comment]
This reminds me of the idea to rebuild the entire hotkey system to allow any event source to bind to any action.

**Comment by WizardCM** [comment]
I think alongside some sort of indicator in the rows that contain duplicate items, I'd also show a warning text at the bottom of the window similar to how we do for surround audio or MP4 recording. And some sort of method to filter by duplicates to make it easier to clear them out.

**Comment by dodgepong** [comment]
This bounty has been claimed and completed. https://github.com/obsproject/obs-studio/pull/5297
